controller/share_manager_controller.go

```diff
@@ -1050,6 +1045,92 @@ func (c *ShareManagerController) getShareManagerTolerationsFromStorageClass(sc *
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to get setting value %v", types.SettingNameStorageNetworkForRWXVolumeEnabled)
+	}
+
+	return types.IsStorageNetworkForRWXVolume(storageNetwork, storageNetworkForRWXVolumeEnabled), nil
+}
+
+func (c *ShareManagerController) checkStorageNetworkApplied() (bool, error) {
+	targetSettings := []types.SettingName{types.SettingNameStorageNetwork, types.SettingNameStorageNetworkForRWXVolumeEnabled}
+	for _, item := range targetSettings {
+		if applied, err := c.ds.GetSettingApplied(item); err != nil || !applied {
+			return applied, err
+		}
+	}
+	return true, nil
+}
+
+func (c *ShareManagerController) canCleanupService(ns, shareManagerName string) (bool, error) {
+	service, err := c.ds.GetService(c.namespace, shareManagerName)
+	if err != nil {
+		// if NotFound, means the service/endpoint is already cleaned up
+		// The service and endpoint are related with the kubernetes endpoint controller.
+		// It means once the service is deleted, the corresponding endpoint will be deleted automatically.
+		if apierrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, errors.Wrap(err, "failed to get service")
+	}
+
+	// check the settings status of storage network and storage network for RWX volume
+	settingsApplied, err := c.checkStorageNetworkApplied()
+	if err != nil {
+		return false, errors.Wrap(err, "failed to check if the storage network settings are applied")
+	}
+	if !settingsApplied {
+		c.logger.Warn("Storage network settings are not applied, do nothing")
+		return false, nil
+	}
+
+	storageNetworkForRWXVolume, err := c.isStorageNetworkForRWXVolume()
+	if err != nil {
+		return false, err
+	}
+
+	// no need to cleanup because looks the service file is correct
+	if storageNetworkForRWXVolume {
+		if service.Spec.ClusterIP == core.ClusterIPNone {
+			return false, nil
+		}
+	} else {
+		if service.Spec.ClusterIP != core.ClusterIPNone {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func (c *ShareManagerController) cleanupService(shareManager *longhorn.ShareManager) error {
+	if ok, err := c.canCleanupService(c.namespace, shareManager.Name); !ok || err != nil {
+		if err != nil {
+			return errors.Wrapf(err, "failed to check if we can cleanup service and endpoint for share manager %v", shareManager.Name)
+		}
+		return nil
+	}
+
+	// let's cleanup
+	c.logger.Infof("Deleting Service for share manager %v", shareManager.Name)
+	err := c.ds.DeleteService(c.namespace, shareManager.Name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return errors.Wrapf(err, "failed to delete service for share manager %v", shareManager.Name)
+	}
+
+	// we don't need to cleanup the endpoint because the kubernetes endpoints_controller
+	// will sync the service then clean up the corresponding endpoint.
+	// https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpoint/endpoints_controller.go#L374-L392
```

Comment:
Should avoid using the master branch as the targeted lines may change over time.

Reply:
Nice suggestion! Updated!